### PR TITLE
example Pipeline definition and cad cli skeleton

### PR DESCRIPTION
**Pull request by ramonbutter**

### CODE REVIEW

File: deploy/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+```console
+oc apply -f . 
+```
+
+**Note**: The pipeline require a persistant storage. The pvc defined here only works for AWS, so for local testing on a crc a pvc with the same name should be created manually.
+
+The CRs are going to be created in the `configuration-anomaly-detection` namespace.
+
+After applying the CRs, a Weblistner will be opened for triggereing pipelines. F.e. http://el-pipeline-event-listener.ci.svc.cluster.local:8080 on CRC.
```
Comment (Tessg22) [inline deploy/README.md:25]: 'Weblistner' and 'triggereing' typos

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "cad",
+	Short: "A util of configuration-anomaly-detection (CAD) checks",
+	Long: `A longer description that spans multiple lines and likely contains
+examples and usage of using your application. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
```
Comment (georgettica) [inline cmd/root.go:38]: remove this if it's not used?

File: deploy/pvc-cad.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: cad-pvc
```
Comment (georgettica) [inline deploy/pvc-cad.yaml:4]: does that mean that only one run can happen in parralel ? (only one PVC can be bound to one pod

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,26 @@
+module configuration-anomaly-detection
```
Comment (georgettica) [inline go.mod:1]: shoudn't this be 
```suggestion
module github.com/openshift/configuration-anomaly-detection
```

File: main.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import "cad/cmd"
```
Comment (georgettica) [inline main.go:18]:
```suggestion
import "github.com/openshift/configuration-anomaly-detection/cmd"
```

**Review by georgettica — COMMENTED** [review]
went through everything aside the README.md

feel free to look and comment accordingly

File: cmd/root.go (hunk 2 above)
Comment (georgettica) [reply cmd/root.go:38]: this is the only place the short description is updated. feel free not to change this

**Comment by georgettica** [comment]
a next step would be to migrate the configs to kustomize (makes setting the namespace much easier, plus additional parameters

File: cmd/root.go (hunk 2 above)
Comment (ramonbutter) [reply cmd/root.go:38]: good point. I would say lets add a proper description

File: deploy/pvc-cad.yaml (hunk 3 above)
Comment (ramonbutter) [reply deploy/pvc-cad.yaml:4]: Generally, mounting a PV to several pods shouldn't be an issue. But yes, we have to change the accessMode. 
As we discussed, the idea with the PV was to have a persistent way of storing logs for debugging. As execution information is lost, once the pod is deleted. But maybe there is a tektonish way to solve that without PVC. I am not sure right now.

**Comment by georgettica** [comment]
Not sure if this should stay open until all data from it will be written down

File: deploy/serviceaccount.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: cad-eventlistener-rolebinding
+  namespace: configuration-anomaly-detection
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: cad-eventlistener-role
+  namespace: cluster-anomaly-detection
```
Comment (georgettica) [inline deploy/serviceaccount.yaml:60]: changed to `configuration-anomaly-detection`, breaks the builder image

File: deploy/README.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+First, apply the subscription to the pipeline operator:
+
+```console
+oc apply -f pipeline-operator-subscription.yaml
+```
+
+Wait a minute until it becomes available, then apply the rest:
+
+```console
+oc apply -f .
```
Comment (georgettica) [inline deploy/README.md:18]: add a section to verify the `el-pipeline-event-listener` works as expected

File: deploy/task-cad-checks.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      type: string
+      description: Json string of the event data
+  results:
+    - name: configuration-anomaly
+      description: Cloud infrastructure anomaly type (f.e. None, HostsDown, ...)
+    - name: customer-caused
+      description: Configuration anomaly caused by customer detected
+  steps:
+    - name: check-infrastructure
+      image: quay.io/app-sre/cad-cli:latest
```
Comment (georgettica) [inline deploy/task-cad-checks.yaml:18]: this image needs to be promoted before the PR is merged (or at least to be noted that it is required

**Review by georgettica — COMMENTED** [review]
found during testing a different PR

File: deploy/task-cad-checks.yaml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      command: ["/bin/bash", "-c"]
+      args:
+        - |-
+          RESULT=$(cad test-infra --event '$(params.event)'
```
Comment (georgettica) [inline deploy/task-cad-checks.yaml:22]:
```suggestion
          RESULT=$(cad test-infra --event '$(params.event)')
```
typo in command

File: deploy/serviceaccount.yaml [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+metadata:
+  creationTimestamp: null
+  name: cad-eventlistener-clusterrole
+rules:
+  - apiGroups:
+      - triggers.tekton.dev
+    resources:
+      - clusterinterceptors
+    verbs:
+      - list
```
Comment (georgettica) [inline deploy/serviceaccount.yaml:48]:
```suggestion
      - list
      - watch
```
failed by getting the error:
```
E0221 13:10:41.885069       1 reflector.go:138] github.com/tektoncd/triggers/pkg/client/informers/externalversions/factory.go:117: Failed to watch *v1alpha1.ClusterInterceptor: unknown (get clusterinterceptors.triggers.tekton.dev)
```

**Comment by georgettica** [comment]
committed my changes but didn't create a PR
https://github.com/georgettica/configuration-anomaly-detection/tree/georgettica/fix-tekton-pipeline

hope this helps

File: cmd/root.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		// Search config in home directory with name ".cad" (without extension).
+		viper.AddConfigPath(home)
+		viper.SetConfigType("yaml")
+		viper.SetConfigName(".cad")
+	}
+
+	viper.AutomaticEnv() // read in environment variables that match
+
+	// If a config file is found, read it in.
+	if err := viper.ReadInConfig(); err == nil {
```
Comment (shibumi) [inline cmd/root.go:83]: What happens if a file cannot be found? and why do we printing a "successful" log to stderr?

**Comment by georgettica** [comment]
@shibumi this PR won't be merged.
the efforts will continue in a different PR (which I have not submitted as of yet

**Comment by shibumi** [comment]
Closing this after speaking with @ramonbutter  about it.

**Pull request closed without merging**